File: infrastructure/servers/patient_server.py
import sys, os

# Agregar la raiz del proyecto al path
ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
sys.path.insert(0, ROOT)

import uvicorn
from fastapi import FastAPI
from importlib import import_module

# Importar modulos (se usa importlib porque "in" es palabra reservada en Python)
patient_repo_mod = import_module("infrastructure.adapters.out.in_memory_patient_repository")
patient_svc_mod = import_module("Application.patient_service")
patient_ctrl_mod = import_module("infrastructure.adapters.in.patient_controller")

InMemoryPatientRepository = patient_repo_mod.InMemoryPatientRepository
PatientService = patient_svc_mod.PatientService
create_patient_router = patient_ctrl_mod.create_patient_router

# Se conectan las capas: repositorio -> servicio -> controlador
patient_repository = InMemoryPatientRepository()
patient_service = PatientService(patient_repository)
patient_router = create_patient_router(patient_service)

app = FastAPI(title="Servicio de Pacientes", description="Examen Taller 6 usando Arquitectura Hexagonal")
app.include_router(patient_router)

if __name__ == "__main__":
    print("Servicio de Pacientes corriendo en http://localhost:8001")
    print("Swagger UI en http://localhost:8001/docs")
    uvicorn.run(app, host="0.0.0.0", port=8001)

File: infrastructure/servers/patientServer.js
const express = require('express');
const swaggerUi = require('swagger-ui-express');
const InMemoryPatientRepository = require('../adapters/out/InMemoryPatientRepository');
const PatientService = require('../../Application/PatientService');
const createPatientRouter = require('../adapters/in/PatientController');

// Se conectan las capas: repositorio -> servicio -> controlador
const patientRepository = new InMemoryPatientRepository();
const patientService = new PatientService(patientRepository);
const patientRouter = createPatientRouter(patientService);

// Documentacion Swagger para la interfaz /docs
const swaggerDocument = {
    openapi: '3.0.0',
    info: {
        title: 'Servicio de Pacientes',
        description: 'API REST para gestionar pacientes - Arquitectura Hexagonal',
        version: '1.0.0'
    },
    servers: [{ url: 'http://localhost:8001' }],
    paths: {
        '/patients': {
            get: {
                summary: 'Consultar todos los pacientes',
                tags: ['Pacientes'],
                responses: {
                    200: { description: 'Lista de pacientes' }
                }
            },
            post: {
                summary: 'Registrar un paciente',
                tags: ['Pacientes'],
                requestBody: {
                    required: true,
                    content: {
                        'application/json': {
                            schema: {
                                type: 'object',
                                required: ['nombre', 'email'],
                                properties: {
                                    nombre: { type: 'string', example: 'Juan Perez' },
                                    email: { type: 'string', example: 'juan@mail.com' }
                                }
                            }
                        }
                    }
                },
                responses: {
                    201: { description: 'Paciente registrado' }
                }
            }
        },
        '/patients/{id}': {
            get: {
                summary: 'Consultar paciente por ID',
                tags: ['Pacientes'],
                parameters: [
                    { name: 'id', in: 'path', required: true, schema: { type: 'integer' } }
                ],
                responses: {
                    200: { description: 'Paciente encontrado' },
                    404: { description: 'Paciente no encontrado' }
                }
            },
            delete: {
                summary: 'Eliminar un paciente',
                tags: ['Pacientes'],
                parameters: [
                    { name: 'id', in: 'path', required: true, schema: { type: 'integer' } }
                ],
                responses: {
                    200: { description: 'Paciente eliminado' },
                    404: { description: 'Paciente no encontrado' }
                }
            }
        }
    }
};

const app = express();
app.use(express.json());
app.use('/docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
app.use('/patients', patientRouter);

const PORT = 8001;
app.listen(PORT, () => {
    console.log(`Servicio de Pacientes corriendo en http://localhost:${PORT}`);
    console.log(`Swagger UI en http://localhost:${PORT}/docs`);
});
